src/main/java/core/basesyntax/service/RegistrationServiceImpl.java

```diff
@@ -5,10 +5,73 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MIN_LOGIN_LENGTH = 6;
+    private static final int MIN_PASSWORD_LENGTH = 6;
+    private static final int MIN_USER_AGE = 18;
+
     private final StorageDao storageDao = new StorageDaoImpl();
 
     @Override
     public User register(User user) {
-        return null;
+        validateUser(user);
+
+        return storageDao.add(user);
+    }
+
+    private void validateLoginUniqueness(String login) {
+        if (storageDao.get(login) != null) {
+            throw new UserDataInvalidException("User with such login already exists");
+        }
+    }
+
+    private void validateLogin(String login) {
+        if (login.length() < MIN_LOGIN_LENGTH) {
+            throw new UserDataInvalidException("User`s login length should contain at least"
+                    + MIN_LOGIN_LENGTH
+                    + "characters, got login with " + login.length() + "characters");
+        }
+    }
+
+    private void validateAge(Integer age) {
+        if (age < MIN_USER_AGE) {
+            throw new UserDataInvalidException("User should be at least "
+                    + MIN_USER_AGE
+                    + " years old, got " + age + "years");
+        }
+    }
+
+    private void validatePassword(String password) {
+        if (password.length() < MIN_PASSWORD_LENGTH) {
+            throw new UserDataInvalidException("User`s password length should contain at least "
+                    + MIN_PASSWORD_LENGTH
+                    + " characters, got password with " + password.length() + "characters");
+        }
+    }
+
+    private void checkNotNull(Object value, String errorMessage) {
+        if (value == null) {
+            throw new UserDataInvalidException(errorMessage);
+        }
+    }
+
+    private void validateUser(User user) {
+        if (user == null) {
+            throw new UserDataInvalidException("User can`t be null");
+        }
+
+        String login = user.getLogin();
+        String password = user.getPassword();
+        Integer age = user.getAge();
+
+        checkNotNull(login, "User`s login can`t be null");
```

Comment:
replace this logic to validate.... methods

Reply:
Changed

Reply:
The idea here was to not call methods without need.
For example:
String login = `properValue`;
Integer age = `properValue`;
String password = `null`;

validateAge(age); // Will be executed without error
validateLogin(login); // Will be executed without error
validateLoginUniqueness(login); // Will be executed without error
validatePassword(password); // Will throw error, registration is not success, but all previous methods were executed without need.

And it realy doesn't matter here, since we are not performing any resource-intensive operations inside of those methods